frame/election-providers/src/onchain.rs

```diff
@@ -0,0 +1,190 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of [`sp_election_providers::ElectionProvider`] that does an on-chain
+//! sequential phragmen.
+
+use sp_arithmetic::PerThing;
+use sp_election_providers::{ElectionDataProvider, ElectionProvider};
+use sp_npos_elections::{
+	ElectionResult, ExtendedBalance, IdentifierT, PerThing128, Supports, VoteWeight,
+};
+use sp_runtime::RuntimeDebug;
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData, prelude::*};
+
+/// Errors of the on-chain election.
+#[derive(RuntimeDebug, Eq, PartialEq)]
+pub enum Error {
+	/// An internal error in the NPoS elections crate.
+	NposElections(sp_npos_elections::Error),
+	/// An assignment failed to pass the feasibility check
+	Feasibility(&'static str),
+}
+
+impl From<sp_npos_elections::Error> for Error {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		Error::NposElections(e)
+	}
+}
+
+/// A simple on-chian implementation of the election provider trait.
+///
+/// This will accept voting data on the fly and produce the results immediately.
+///
+/// ### Warning
+///
+/// This can be very expensive to run frequently on-chain. Use with care.
+pub struct OnChainSequentialPhragmen<T: Config>(PhantomData<T>);
+
+/// Configuration trait of [`OnChainSequentialPhragmen`].
+///
+/// Note that this is similar to a pallet traits, but [`OnChainSequentialPhragmen`] is not a pallet.
+pub trait Config {
+	/// The account identifier type.
+	type AccountId: IdentifierT;
+	/// The block number type.
+	type BlockNumber;
+	/// The accuracy used to compute the election:
+	type Accuracy: PerThing128;
+	/// Something that provides the data for election.
+	type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+}
+
+impl<T: Config> ElectionProvider<T::AccountId, T::BlockNumber> for OnChainSequentialPhragmen<T>
+where
+	ExtendedBalance: From<<T::Accuracy as PerThing>::Inner>,
+{
+	type Error = Error;
+	type DataProvider = T::DataProvider;
+
+	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
+		let voters = Self::DataProvider::voters();
+		let targets = Self::DataProvider::targets();
+		let desired_targets = Self::DataProvider::desired_targets() as usize;
+
+		let mut stake_map: BTreeMap<T::AccountId, VoteWeight> = BTreeMap::new();
+
+		voters.iter().for_each(|(v, s, _)| {
+			stake_map.insert(v.clone(), *s);
+		});
+
+		let stake_of = Box::new(|w: &T::AccountId| -> VoteWeight {
+			stake_map.get(w).cloned().unwrap_or_default()
+		});
+
+		let ElectionResult {
+			winners,
+			assignments,
+		} = sp_npos_elections::seq_phragmen::<_, T::Accuracy>(desired_targets, targets, voters, None)
+			.map_err(Error::from)?;
+
+		// check all assignments for feasibility, based on election data provider.
+		assignments
```

Comment:
This is super conservative: The whole point of this is to deal with the recent slashes and ensure that they are not included in the solution, mainly in external solutions. Ideally, an election data provider should differentiate between the two and in the case of an on-chain election, simply trust the data.

Reply:
Given the comment here, this comment is on the right track and this feasibility_check by the data provider is basically redundant after all. 

https://github.com/paritytech/substrate/pull/7319#discussion_r553535068